Implemented GuiIcon, GuiSeekBar and ValueFormatter + code cleanup and optimization

SrRapero720 opened the pull request:
i add the same features of #196 except the font scaling feature, which needs to do more deep testing because in 1.20.1 scale the pose stack not longer works. So, i will add that feature in another PR when i discover how to make it properly work

now, this PR:
- Adds ValueFormatter (old called ValueDisplay) for the fancy text parsing (angle, block counts, pixels, etc)
- Adds GuiSeekBar (a GuiSlider but with a different style)
- Adds GuiIcon (renames 3 classes to make it fit)
- Performs a huge analyzeusing intelliJ analyzer tool and the code i noted while i do the backport
  - Basically i fix minor performance issues, autoreferenced methods (which ends in a stackoverflowerror), if chains to switchs, fix minor bugs (such as check if block is null instead of is air), code modernization, simplification of lambas (super expanded consumer to fire a method which requires a consumer), redundant checks (null or positive was already checked), some aritmetic operations are 0 because it are casted to int, and many other things...
- Imports optimizations (most irrelevant and i just use the IntelliJ automatic tool to do that)


I made the code optimization in a separated commit to make more easier read the changes. i add some "FIXME" because some chunks of code had confusing auras (And i am doing this at 3am so...)

CreativeMD committed:
Bumped version to 2.11.24

SrRapero720 committed:
Merge branch 'CreativeMD:1.20' into 1.20

SrRapero720 committed:
fix project refuses to build by missing Curseforge key

SrRapero720 committed:
added value formatter for GuiSlider

SrRapero720 committed:
Merge remote-tracking branch 'origin/1.20' into 1.20

SrRapero720 committed:
added GuiSeekBar

SrRapero720 committed:
fix suggestion is always rendered when is defined

SrRapero720 committed:
added getters for private fields

SrRapero720 committed:
general code optimizations

SrRapero720 committed:
general imports optimizations

SrRapero720 committed:
switch to fabric mixin annotation processor (Fixes issues injecting lambdas)

SrRapero720 committed:
added GuiIcon widget

- Renames old GuiIcon class to GuiTexIcon
- Renames GuiIconButton to GuiButtonIcon (for consistency)

SrRapero720 commented:
IDK from where commes the bumped version commit, somehow it appears

CreativeMD commented on src/main/java/team/creative/creativecore/common/config/group/Usergroup.java:
```diff
@@ -17,8 +18,7 @@ public Usergroup() {
     public Usergroup(PlayerSelector... selectors) {
-        for (PlayerSelector selector : selectors)
-            filters.add(selector);
+        filters.addAll(Arrays.asList(selectors));
```
I will rather leave it as before, this is just slower and a for loop is fine as well. Just a small thing, but prefer to keep it as it was.

SrRapero720 replied:
i do the change by performance reasons, i check again the check (after sleep 12 hours) and i see performance is worse. mosly because it copies the array 3 times. for some reason i ignore the fact it is on constructor, so i change the add all to ``Array.asList()`` which literally creates a new ArrayList instance using the same argumented array.

resulting code looks like this and is much better than before
![image](https://github.com/CreativeMD/CreativeCore/assets/29090346/8449121e-5217-464d-bb89-43bda530ae02)

CreativeMD commented on src/main/java/team/creative/creativecore/common/gui/controls/collection/GuiStackSelector.java:
```diff
@@ -259,15 +258,17 @@ public static class GuiBlockSelector extends SearchSelector {
-                return Block.byItem(stack.getItem()) != null && !(Block.byItem(stack.getItem()) instanceof AirBlock);
+            if (super.allow(stack)) {
+                Block.byItem(stack.getItem());
+                return !(Block.byItem(stack.getItem()) instanceof AirBlock);
```
Something went wrong here I guess?

SrRapero720 replied:
_my brain at some point collapses and adds another Block.byItem for no reason_
i remove the nullcheck because result was never null, `instanceof` checks also checks if it wasn't null. and null is not an instance of null

the extra Block.byItem line removed too
![image](https://github.com/CreativeMD/CreativeCore/assets/29090346/71d46042-7350-47d9-a89c-a386f15cd39c)

CreativeMD commented on src/main/java/team/creative/creativecore/common/gui/controls/simple/GuiSeekBar.java:
```diff
@@ -0,0 +1,61 @@
+            this.maxValue = valueSuppliers.getMax();
+            this.value = valueSuppliers.getValue();
+        }
+        double percent = this.getPercentage();
```
Rather do a super call here, because I do not see any difference to the default behavior.

SrRapero720 replied:
it changes the start position of the thumb
making it look like this

i have to say, i didn't port the code to 1.20.1, and i didn't see the new "ProgressBar" ControlFormatting, which fullfits the behavior of the current seekbar, code can be easier simplified
![image](https://github.com/CreativeMD/CreativeCore/assets/29090346/738a55db-e8ff-4ddf-8395-add698412ffd)

SrRapero720 replied:
i simplified the Seekbar using Longsupplier (begin the used primitive for time seeking) and a consumer to fire value updates.

Now it internaly looks like the child of GuiProgressbar and GuiSlider
![image](https://github.com/CreativeMD/CreativeCore/assets/29090346/6279760f-4e58-4e33-8401-a034f3c105d7)

![image](https://github.com/CreativeMD/CreativeCore/assets/29090346/384d6c94-8348-4ba1-99c9-f3ce86193dfa)

CreativeMD commented on src/main/java/team/creative/creativecore/common/gui/controls/tree/GuiTreeDragPosition.java:
```diff
@@ -1,34 +1,30 @@
+public record GuiTreeDragPosition(
+        team.creative.creativecore.common.gui.controls.tree.GuiTreeDragPosition.ItemPosition position,
+        GuiChildControl child, GuiTreeItem item, GuiChildControl above) {
+
```
Actually in that case a class with final fields seems more convenient, that way we avoid all these method calls in `GuiTree`.

SrRapero720 replied:
Recods are optimized at JVM level, making more strict optimizations to do a better job managing data. A really noticeable optimizations with a medium-large number of instances

CreativeMD replied:
Try, but in this case performance does not matter at all. I mean we can also include it, just not a fan of these brackets I have to use every time I call it's fields. So I only do that of there is an actual benefit.

SrRapero720 replied:
I can revert it if you want it. But for what was used is more convenient have it as record
also doing some test, even on 10 size records wins with a 10-15% of enhancement

CreativeMD commented on src/main/java/team/creative/creativecore/common/gui/event/GuiEventManager.java:
```diff
@@ -6,10 +6,10 @@
     
-    private List<GuiEventHandler> handlers = new ArrayList<>();
+    private final List<GuiEventHandler<?>> handlers = new ArrayList<>();
     
```
Probably your formatter ... I personally don't see any advantage in adding this brackets. I would rather leave them out.

SrRapero720 replied:
Brackets are really important to cast values
Is not the same use List#add to List<String>#add. Brackets makes implicit cast to class arguments.

Is a good practice having it just for code sanity

CreativeMD replied:
Not sure if I get your point. The compiled code is exactly the same. I find it quite annoying to work with and it makes the code more difficult to read.

SrRapero720 replied:
adding brackets forces you to cast whatever was the class arg

Without bracket basically you list is an `List<Object>`, but using `List<?>` mean your list type argument is a mistery and needs codder criteria to cast it back

so if you get and item from a List and returns type ``GuiEventHandler<ClassA>`` but you cast to `GuiEventHandler<ClassB>` and ``ClassB`` doesn't extends from ``ClassA`` that basically leads in a ``ClassCastException``

so, in some situations compiler forces you to do manual check for whatever type was when you attemped to use the object

CreativeMD replied:
All generics are lost during runtime. Furthermore the result in code is the same. It's literally the same at least in eclipse.

CreativeMD commented on src/main/java/team/creative/creativecore/common/gui/packet/ControlSyncPacket.java:
```diff
@@ -24,8 +24,6 @@ public ControlSyncPacket() {}
     public void execute(Player player, IGuiIntegratedParent container) {
         GuiSyncControl sync = GuiSyncHolder.followPathControl(path, container);
-        if (sync == null)
-            throw new RuntimeException("Could not find sync " + path);
```
Would remove it? This can actually happen.

SrRapero720 replied:
Yes, because GuiSyncHolder.followPathControl already do the same check, which makes that check redundant and never fired

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/math/box/BoxFace.java:
```diff
@@ -113,7 +129,7 @@ public Boolean isFacingOutwards(boolean first, boolean inverted, Vec3f normal) {
         if (outwardOne != null && outwardTwo == null)
             return outwardOne;
-        if (outwardOne == null && outwardTwo != null)
+        if (outwardOne == null)
```
It is actually important that outwardTwo is not null.

SrRapero720 replied:
At this point of the checks, outwardTwo is not null
Because if first check was not fulfilled (outwardOne is null and outwardTwo is not null), the second checks what check was not fulfilled (outwardOne is null), making implicit outwardTwo is not null



Basically, the not null check was already done and the second check is just redundant

SrRapero720 replied:
**First Check: `if (outwardOne == outwardTwo)`**
- At this stage, we not only verify if `outwardOne` equals `outwardTwo`, but also check if `null == null`. If they don't match, it indicates that one of them is not null.

**Second Check: `if (outwardOne != null && outwardTwo == null)`**
- If they don't match, it implies a difference in value. Failure to meet this condition suggests that either `outwardOne` is null or `outwardTwo` is not null.

**Third Check: `if (outwardOne == null && outwardTwo != null)`**
- As previously mentioned, the comparison `if (outwardOne == outwardTwo)` also encompasses the scenario `null == null`. Therefore, if `outwardOne` is null, `outwardTwo` cannot be null, as the first condition would have been met, rendering the null check for `outwardTwo` redundant.

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/math/box/BoxUtils.java:
```diff
@@ -64,7 +64,7 @@ private static void includeMaxRotationInBox(ABB box, Vec3d vec, Axis axis, doubl
         if (rotation >= 90) {
             while (quarterRotation <= Math.abs(rotation) && quarterRotation < 360) {
-                Facing facing = positive ? state.clockwiseMaxFacing() : state.counterMaxClockwiseFacing();
+                Facing facing = state.clockwiseMaxFacing();
```
Pretty sure this does not produce the same results.

SrRapero720 replied:
positive check is rotation > 0 and second check do the same but for 90, which is positive, and makes redundant check again if rotation is again positive

SrRapero720 replied:
yep, in the ``if check context`` we already do if rotation is positive, so there is no reason to check it again

![image](https://github.com/CreativeMD/CreativeCore/assets/29090346/9b7c1460-203d-4c79-90ed-78064d35c3b8)

CreativeMD replied:
Rotation can be negative. In that case Positive is false.

SrRapero720 replied:
ok i read the change again and now i get your point

SrRapero720 replied:
i don't get it.

CreativeMD replied:
Just revert to old.

SrRapero720 replied:
Inside the context of `rotation => 90` it can't be negative, never. Because rotation is over positive 90

The `positive ? :` check was done inside `rotation => 90` which already checks if rotation is positive (because rotation is above 0, even above 89)

Also just mentioned, rotation is never modified, quarterRotation is

SrRapero720 replied:
I will revert it back, but is a redundand check in that context

SrRapero720 replied:
done

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/math/geo/VectorFan.java:
```diff
@@ -955,31 +946,27 @@ else if (allValue)
         
         if (left.size() >= 3 && done != null)
-            done.add(new VectorFan(left.toArray(new Vec3f[left.size()])));
+            done.add(new VectorFan(left.toArray(new Vec3f[0])));
```
new Vec3f[0] will result in another array being created to match the size of the list.

SrRapero720 replied:
That's on purpose, the new array instance is not used and only required to cast array to the proper Class. Make the array with a pre established size wastes memory.

SrRapero720 replied:
ignore my stupidity, i will rever this damn change. that makes performance worse :P

SrRapero720 replied:
https://www.baeldung.com/java-collection-toarray-methods#benchmarks-on-newer-jdks

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/math/geo/VectorFan.java:
```diff
@@ -955,31 +946,27 @@ else if (allValue)
         
         if (right.size() < 3)
             return null;
-        return new VectorFan(right.toArray(new Vec3f[right.size()]));
```
same as the one before.

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/type/list/CopyArrayCollection.java:
```diff
@@ -73,7 +73,7 @@ public CopyArrayCollection(Collection<? extends E> c) {
             content = Arrays.copyOf(((CopyArrayCollection) c).content, ((CopyArrayCollection) c).content.length, Object[].class);
         else
-            content = c.toArray(new Object[c.size()]);
+            content = c.toArray(new Object[0]);
```
Again, this creates an empty array for nothing. Rather leave it as it is.

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/type/list/Pair.java:
```diff
@@ -36,7 +36,7 @@ public int hashCode() {
     public boolean is(K key) {
         if (this.key != null)
             return this.key.equals(key);
-        return this.key == key;
```
This basically checks if the other key is null. Don't see a need to change it, of course you can go with == null as well.

SrRapero720 replied:
That's for code readbility

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/type/list/SortingList.java:
```diff
@@ -192,7 +190,7 @@ public boolean contains(Object arg0) {
     @Override
     public boolean containsAll(Collection<?> arg0) {
-        return entries.containsAll(arg0);
+        return new HashSet<>(entries).containsAll(arg0);
```
I don't see why you would do that?

SrRapero720 replied:
HashSet is more faster to run containsAll()

SrRapero720 replied:
I think it can be enhanced to use it as storage and not scare the GC every containsAll call. L

CreativeMD replied:
I don't think this is faster ... you create a new HashSet every time you call contains all. This means all entries need to be hashed and put into buckets. If you have performance issues at least cash the hashset, then performance might be improved.

SrRapero720 replied:
reverted, i want to do some benchsmarks arround Sets and arrays

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/type/list/Tuple.java:
```diff
@@ -36,7 +36,7 @@ public int hashCode() {
         if (this.key != null)
             return this.key.equals(key);
-        return this.key == key;
+        return key == null; // FIXME: uhh
```
Same as the one before in `Pair`

SrRapero720 replied:
I think it can stay just for code readbility

SrRapero720 replied:
But I want to ask what behavior is the required one, if return false even if both are null or keep the current one to validate even it null == null?

CreativeMD replied:
Sure thing, but would be cool if you can remove the // FIXME comment (also for the `Pair` class).

SrRapero720 replied:
done

CreativeMD commented on build.gradle:
I don't think these changes should be in here.

SrRapero720 replied:
The only change I accidentally commit was the removal of diff plugin. Other changes are on purpose, fabric annotation processor works with vanilla mixins and solves issues Injecting code to anonymous classes. (Mappings issues)

Again, works fine with forge because it is an annotation processor, which runs on compiling

CreativeMD replied:
Yeah ... would be cool if you could revert these changes, because I don't want to change this file in that way. There are plenty of changes for 1.20.4 already.

SrRapero720 replied:
the 1.20.4 changes correspond to Neoforge which doesn't use mappings. so these changes doesn't apply for 1.20.4
anyway, checking mixins i think CC can survive without fabric annotationProcessor.

not sure what's your plan but is reverted [Except the CurseForge fix because it can't build without it (because key is null)]

CreativeMD commented on src/main/java/team/creative/creativecore/common/config/gui/ClientSyncGuiLayer.java:
```diff
@@ -112,14 +112,15 @@ public void load(CheckTree<ConfigKey>.CheckTreeEntry entry) {
-                second.add(new GuiLabel(caption).setTitle(Component.literal(caption)).setTooltip(new TextBuilder().translateIfCan(comment).build()));
+            if (key.content != null) {
+                // FIXME: is normal use holder.path on a string concatenation?
+                String caption = translateOrDefault("config." + String.join(".", holder.path() + "." + key.content.name + ".name"), key.content.name);
```
I think it should be
String caption = translateOrDefault("config." + String.join(".", holder.path()) + "." + key.content.name + ".name", key.content.name);

SrRapero720 replied:
holder.path() returns an String array.

CreativeMD replied:
Yep, the current code is actually wrong. That's why I moved the bracket to a different place to make it work (as shown above).

SrRapero720 replied:
done

CreativeMD reviewed (changes requested):
First of all thanks a lot for your work!

Unfortunately there are quite a lot of formatting changes ... especially imports. Not sure how these happen, but they make it quite a lot more to review than actually necessary.

There a few things I commented on.

SrRapero720 committed:
Revert "switch to fabric mixin annotation processor (Fixes issues injecting lambdas)"

This reverts commit 4674d698f2e2dee60456c556c43bdacb451ccf59.

SrRapero720 committed:
gui control returns null by default instead of the custom tooltip

this had no changes on override methods and GuiControl objects with no instances.

SrRapero720 committed:
fixed bad optimized constructor

SrRapero720 committed:
removed useless code

SrRapero720 committed:
update seekbar: now uses progressbar control formatting

SrRapero720 committed:
use ValueFormatter GuiProgressbar

SrRapero720 committed:
minor changes

SrRapero720 committed:
check if state wasn't null

SrRapero720 committed:
reverted HashSet wrapping

SrRapero720 committed:
remove comments

SrRapero720 committed:
empty

SrRapero720 committed:
reverted fixes

SrRapero720 committed:
minor fixes

SrRapero720 committed:
Merge branch 'CreativeMD:1.20' into 1.20

SrRapero720 committed:
minor changes

SrRapero720 commented:
i answered all review messages, i stay waiting for your answer :D

SrRapero720 committed:
renamed GuiTexIcon to Icon

SrRapero720 committed:
renamed TextFormatter to ValueParser (for congruency with textFormatter in TextFields

SrRapero720 committed:
Revert "general imports optimizations"

This reverts commit 59b1c00b

SrRapero720 committed:
added ValueParser support for GuiSteppedSlider

SrRapero720 committed:
fix for GuiSeekBar

SrRapero720 committed:
rewrite ValueParser

- splitted parsers interfaces per primitive types (ignoring float)
- secure all parser to be "0 / 0" prof
- optimize Component code for block text
- simplified code

SrRapero720 committed:
enhance trycatchs with checks for 0 values

SrRapero720 committed:
fix check for 0 values (again)

CreativeMD commented on build.gradle:
Again these changes should not be in there as it will break my setup. I see that you need these to make it work for you, but you should not commit it to the pull request.

SrRapero720 replied:
i will reverted at the end when all requested changes are finished, just to do test builds more faster

SrRapero720 replied:
done

SrRapero720 committed:
bad return type

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/type/list/SortingList.java:
```diff
@@ -66,8 +62,8 @@ public void setBlacklist() {
     public void addSortingObjects(Object... objects) {
-        for (int i = 0; i < objects.length; i++) {
-            addSortingObject(objects[i]);
+        for (Object object: objects) {
```
Don't see a reason for this change. Many of these changes are just for the sake of it.

SrRapero720 replied:
indeed, there's some parts of the code using it and others no, the change is *for the sake* and for consistency. May not but is better do the change now than never

also, just to note, enhanced for is more faster used in LinkedList. ArrayList and variants performance is the same

CreativeMD replied:
I do that a lot, so the original for loop the preferred way (it also works without creating another object).

SrRapero720 replied:
So I revert it?

CreativeMD commented on src/main/java/team/creative/creativecore/common/util/type/list/Tuple.java:
```diff
@@ -36,12 +36,12 @@ public int hashCode() {
     @Override
     public boolean equals(Object obj) {
-        if (obj instanceof Tuple tuple)
+        if (obj instanceof Tuple<?, ?> tuple)
```
Please avoid these. That's the style of code of this project and it does not make sense to do it different in a few locations.

SrRapero720 replied:
done

SrRapero720 committed:
Minor code changes

SrRapero720 committed:
No brakets

SrRapero720 committed:
revert removal of redundant check

SrRapero720 committed:
revert any gradle change

SrRapero720 commented:
i adressed everything at the moment, except the forloop -> foreach thing which i have no idea what to do
again, is just for the sake, there's no difference, except when is used with HashMaps, which Enhanced for (or for-each) is more faster than normal for loop.

SrRapero720 committed:
expose GuiParent of plus/minus buttons

SrRapero720 committed:
added helpers methods

SrRapero720 committed:
ensure casting

SrRapero720 committed:
Revert "ensure casting"

This reverts commit 4729d8591dbdbcfe7c8e201fd5108e0f8f92fb2a.

SrRapero720 committed:
move button creation to a method

CreativeMD commented:
Thanks a lot for your work! I still don't get why you changed so many for loops. I will revert some of them back. Anyway, nice work! I think there were plenty of improvements. Thanks for spending the time!
